Orbits are now a subsystem

MrsTonedOne opened the pull request:
:cl:
tweak: Orbits now use a subsystem.
tweak: Orbits now bind to object's moving to allow for quicker updates, subsystem fallback for when that doesn't work (for things inside of things or things that move without announcing it)
/:cl:

We bind to Moved() and use datums and lists to track the orbits, no more sleeps, no more delay. (When Moved() works, it's not called for loc setting, and if a thing is inside another thing that is moving)
If Moved() isn't getting called, we still run once every 2 ticks to check (old code was running once a tick, so still a massive improvement on processing cost.)

Moved orbit code to a module to save the poor unsorted file from some bloat.

Not using a sleep() per orbiting thing will massively reduce overhead of orbits. I had 3000 orbits running before the mc started throttling the orbit subsystem.

Having it be in the mc will massively reduce the amount of delay added to orbits from lag, ticker subsystems are given a bit or priority, and mc stuff is managed better with respects to lag.

MrsTonedOne committed:
Sleepless perfect orbits #MOGA
We bind to Moved() and use datums and lists to track the orbits, no more sleeps, no more delay.

MrsTonedOne committed:
Adds some null checks to orbit checks

MrsTonedOne committed:
Forget to set orbiting.
Also sets orbiters before orbiting, to avoid edge cases of deleting a list then recreating it.

MrPerson commented on code/modules/orbit/orbit.dm:
```diff
@@ -0,0 +1,103 @@
+	SpinAnimation(rotation_speed, -1, clockwise, rotation_segments)
+
+	//we stack the orbits up client side, so we can assign this back to normal server side without it breaking the orbit
+	transform = initial_transform
```
This line is both pointless and misleading. It should be removed.

MrStonedOne replied:
No, its needed, other wise when you have two orbits they stack

MrStonedOne replied:
and its not misleading, we do stack them up client side (using animate()) this preserves any existing transform for after the orbits end.

MrPerson replied:
Ah, yeah you're right. I missed that it was taking the initial transform before applying the shift and prerotation. I thought it was just replicating the last animation in SpinAnimation()

PKPenguin321 commented:
Neat

MrStonedOne commented:
There are some unforseen issues with this, I'll have to make it a subsystem (meaning it technically has a sleep)

AdamElTablawy commented:
What impact does orbit not having a delay have?

AdamElTablawy commented:
Also, I don't see something stopping the orbit after you orbit something else unless i've misread something. WHich uh, could be bad if there are stacks and stacks of ghosts orbiting a guy with a spectral sword or something just by orbiting, stopping orbit, and orbiting again, endlessly.

Iamgoofball commented:
Can you add a list of orbiting objects to objects being orbited

MrStonedOne commented:
This already does.

MrStonedOne commented on code/modules/orbit/orbit.dm:
```diff
@@ -0,0 +1,103 @@
+
+	if (orbiter)
+		if (orbiter.orbiting)
+			orbiter.stop_orbit()
```
@AdamElTablawy This line of code stops the existing orbit when you start up a new orbit

MrsTonedOne committed:
Improves orbit, adds subsystem for orbits.
Most orbit loc changes will happen on move, subsystem runs every 2 ticks to make up for when that isn't the case.
